crypto/src/main/kotlin/net/corda/v5/crypto/merkle/MerkleTreeFactory.kt

```diff
@@ -0,0 +1,55 @@
+package net.corda.v5.crypto.merkle
+
+import net.corda.v5.crypto.DigestAlgorithmName
+import net.corda.v5.crypto.SecureHash
+
+/**
+ * [MerkleTreeFactory] creates [MerkleTree]s, [MerkleProof]s and [MerkleTreeHashDigestProvider].
+ */
+interface MerkleTreeFactory {
+    /**
+     * Creates a [MerkleTree]
+     *
+     * @param leaves The leaves of the tree.
+     * @param digestProvider Merkle Tree Hash digest provider used to construct the tree's node and leaf hashes.
+     *
+     * @return A new [MerkleTree] instance.
+     */
+    fun createTree(
+        leaves: List<ByteArray>,
+        digestProvider: MerkleTreeHashDigestProvider
+    ) : MerkleTree
+
+    /**
+     * Recreates a [MerkleProof] from its data content.
+     * Use [net.corda.v5.crypto.merkle.MerkleTree.createAuditProof] to create a proof for a set of leaves for an
+     * existing [MerkleTree].
+     * Use this function to recreate the [MerkleProof] when the [MerkleTree] is not available.
+     *
+     * @param treeSize Number of leaves in the whole tree
+     * @param leaves Leaf items whose inclusion is proved by the proof.
+     * @param hashes The helper hashes needed to reconstruct the whole tree.
+     *
+     * @return A new [MerkleProof] instance.
+     */
+    fun createProof(
```

Comment:
What is the use case for this function? Do we need this on the public API?

Reply:
Hmm. Good point.
I think deserialization will be the way how we will recreate proofs on the verifier sides.
I have removed this from here.

Reply:
Good point.
I think deserialization will be the way how we will recreate proofs on the verifier side.
I have removed it.